cluster/manager.go

```diff
@@ -332,3 +369,126 @@ func (c *ClusterManager) SetPriority(prio int) {
 	nodePriority.Set(prio)
 	c.BroadcastUpdate()
 }
+
+func (c *MemberlistClusterManager) Stop() {
+	c.list.Leave(time.Second)
+}
+
+type SingleNodeManager struct {
+	sync.RWMutex
+	node Node
+}
+
+func (m *SingleNodeManager) Start() {
+	return
+}
+
+func (m *SingleNodeManager) IsPrimary() bool {
+	m.RLock()
+	defer m.RUnlock()
+	return m.node.Primary
+}
+
+func (m *SingleNodeManager) SetPrimary(primary bool) {
+	m.Lock()
+	defer m.Unlock()
+	if m.node.Primary == primary {
+		return
+	}
+	node := m.node
+	node.Primary = primary
+	node.PrimaryChange = time.Now()
+	node.Updated = time.Now()
+	m.node = node
+
+	nodePrimary.Set(primary)
+}
+
+func (m *SingleNodeManager) IsReady() bool {
+	m.RLock()
+	defer m.RUnlock()
+	return m.node.IsReady()
+}
+
+func (m *SingleNodeManager) SetReady() {
+	m.SetState(NodeReady)
+}
+
+func (m *SingleNodeManager) SetReadyIn(t time.Duration) {
+	go func() {
+		// wait for warmupPeriod before marking ourselves
+		// as ready.
+		time.Sleep(t)
+		m.SetReady()
+	}()
+}
+
+func (m *SingleNodeManager) SetState(state NodeState) {
+	m.Lock()
+	defer m.Unlock()
+	if m.node.State == state {
+		return
+	}
+	node := m.node
+	node.State = state
+	node.Updated = time.Now()
+	m.node = node
+	nodeReady.Set(state == NodeReady)
+	return
+}
+
+func (m *SingleNodeManager) ThisNode() Node {
+	m.RLock()
+	defer m.RUnlock()
+	return m.node
+}
+
+func (m *SingleNodeManager) MemberList() []Node {
+	m.RLock()
+	defer m.RUnlock()
+	return []Node{m.node}
+}
+
+func (m *SingleNodeManager) Join(peers []string) (int, error) {
+	//noop
+	return 0, nil
+}
+
+// set the partitions that this node is handling.
+func (m *SingleNodeManager) SetPartitions(part []int32) {
+	m.Lock()
+	defer m.Unlock()
+	node := m.node
+	node.Partitions = part
+	node.Updated = time.Now()
+	m.node = node
+	nodePartitions.Set(len(part))
+	return
+}
+
+// get the partitions that this node is handling.
+func (m *SingleNodeManager) GetPartitions() []int32 {
+	m.RLock()
+	defer m.RUnlock()
+	return m.node.Partitions
+}
+
+// set the priority of this node.
+// lower values == higher priority
+func (m *SingleNodeManager) SetPriority(prio int) {
+	m.Lock()
+	defer m.RUnlock()
```

Comment:
This is first acquiring a write lock and then releasing a read lock. I don't know what happens in this case

Reply:
nothing good.  Will fix it up